File: 프로그래머스/level2/17679.py
def check_block(a, b, board):
  if board[a][b] == board[a][b + 1] == board[a + 1][b]  == board[a + 1][b + 1] :
    return True
  return False

def block_down(M, N, board):
  arr = [[0] * N for _ in range(M)]
  for n in range(N):
    tmp = M - 1
    for m in range(M - 1, -1, -1):
      if board[m][n]:
        arr[tmp][n] = board[m][n]
        tmp -= 1
  return arr

def count(M, N, board):
  result = 0
  for m in range(M):
    for n in range(N):
      if not board[m][n]: 
        result += 1
  return result

def solution(M, N, board):
  board = [list(b) for b in board]
  arr = []
  while True:
    for m in range(M - 1):
      for n in range(N - 1):
        if board[m][n] and check_block(m, n, board):
          arr += [(m, n), (m, n + 1), (m + 1, n), (m + 1, n + 1)]
    if not arr:
      break
    for a in arr:
      board[a[0]][a[1]] = 0
    board = block_down(M, N, board)
    arr = []
  result = count(M, N, board)
  return result

print(solution(6, 6, ['TTTANT', 'RRFACC', 'RRRFCC', 'TRRRAA', 'TTMMMF', 'TMMTTJ']))

File: 프로그래머스/level2/17679.js
const checkBlock = (a, b, arr) => {
  const num = arr[a][b]
  if (num === arr[a + 1][b] && num === arr[a][b + 1] && num === arr[a + 1][b + 1]) {
    return true
  }
  return false
}

const downBlock = (M, N, board) => {
  let arr = Array(M).fill().map(() => Array(N).fill(0))
  for (let j = 0; j < N; j++) {
    let tmp = M - 1
    for (let i = M - 1; i > -1; i-- ) {
      if (board[i][j]) {
        arr[tmp][j] = board[i][j]
        tmp -= 1
      }
    }
  }
  return arr
}

const count = (arr) => {
  let result = 0
  arr.forEach(row => {
    row.forEach(d => {
      if (d === 0) result += 1
    })
  })
  return result
}

const solution = (M, N, board) => {
  board = board.map((v) => [...v])
  let arr
  while (true) {
    arr = []
    for (let i = 0; i < M - 1; i++) {
      for (let j = 0; j < N - 1; j++) {
        if (board[i][j] && checkBlock(i, j, board)) {
          arr = arr.concat([[i, j], [i + 1, j], [i, j + 1], [i + 1, j + 1]])
        }
      } 
    }
    if (!arr.length) {
      break
    }
    arr.forEach((v) => board[v[0]][v[1]] = 0)
    board = downBlock(M, N, board)
  }
  const result = count(board)
  return result
}


console.log(solution(4, 5, ['CCBDE', 'AAADE', 'AAABF', 'CCBBF']))
